weblate/addons/models.py

```diff
@@ -81,15 +92,23 @@ def save(
         self, force_insert=False, force_update=False, using=None, update_fields=None
     ):
         cls = self.addon_class
-        self.project_scope = cls.project_scope
         self.repo_scope = cls.repo_scope
+        self.project_scope = cls.project_scope
+
+        if self.project_scope:
+            self.project = self.component.project
+            self.component = None
 
         # Reallocate to repository
         if self.repo_scope and self.component.linked_component:
             self.component = self.component.linked_component
 
         # Clear add-on cache
-        self.component.drop_addons_cache()
+        if self.component:
+            self.component.drop_addons_cache()
+        elif self.project:
+            for component in self.project.component_set.all():
+                component.drop_addons_cache()
```

Comment:
The cache is in-memory only, so only currently affected objects need to be purged (this avoids possibly expensive loading of all components):

```suggestion
        if self.component:
            self.component.drop_addons_cache()
        if original_component:
            original_component.drop_addons_cache()
```